### PR TITLE
[YUNIKORN-2461] Consider supporting /ws/v1/partition/{partitionName}/queue/{queueName}/applications/:state

### PR DESCRIPTION
**Pull request by blueBlue0102**
### What is this PR for?
This PR continues the work from #861.
After contacting @targetoee and receiving his consent, I am proceeding to complete this issue.

After this PR is merged, I will update the API documentation.

### What type of PR is it?
* [x] - Feature

### What is the Jira issue?
[YUNIKORN-2461](https://issues.apache.org/jira/browse/YUNIKORN-2461)

### How should this be tested?

- Ensure all unit tests pass.
  The CI workflow has passed successfully in my [forked repo](https://github.com/blueBlue0102/yunikorn-core/actions/runs/10303868829)
- Successfully fetch the API locally using Postman.

### Screenshots (if appropriate)
image
image
image
image

### CODE REVIEW

File: pkg/webservice/handlers.go [hunk 1]
```diff
@@ -898,6 +898,56 @@ func getPartitionRules(w http.ResponseWriter, r *http.Request) {
...
+	queueName := vars.ByName("queue")
+	appState := strings.ToLower(vars.ByName("state"))
+	status := strings.ToLower(r.URL.Query().Get("status"))
+
+	queueErr := validateQueue(queueName)
+	if queueErr != nil {
+		buildJSONErrorResponse(w, queueErr.Error(), http.StatusBadRequest)
+		return
+	}
+	partitionContext := schedulerContext.GetPartitionWithoutClusterID(partition)
```
Comment (chia7712) [inline pkg/webservice/handlers.go:919]: `schedulerContext.Load().GetPartitionWithoutClusterID(partition)`

File: pkg/webservice/handlers.go [hunk 2]
```diff
@@ -898,6 +898,56 @@ func getPartitionRules(w http.ResponseWriter, r *http.Request) {
...
+	if queue == nil {
+		buildJSONErrorResponse(w, QueueDoesNotExists, http.StatusNotFound)
+		return
+	}
+	if appState != "active" {
+		buildJSONErrorResponse(w, fmt.Sprintf("The provided state is invalid: %s", appState), http.StatusBadRequest)
+		return
+	}
+	if status != "" && !allowedAppActiveStatuses[status] {
+		buildJSONErrorResponse(w, fmt.Sprintf("The provided status is invalid: %s", status), http.StatusBadRequest)
```
Comment (chia7712) [inline pkg/webservice/handlers.go:934]: `buildJSONErrorResponse(w, allowedActiveStatusMsg, http.StatusBadRequest)`

File: pkg/webservice/handlers.go [hunk 3]
```diff
@@ -898,6 +898,56 @@ func getPartitionRules(w http.ResponseWriter, r *http.Request) {
...
+		buildJSONErrorResponse(w, PartitionDoesNotExists, http.StatusNotFound)
+		return
+	}
+	queue := partitionContext.GetQueue(queueName)
+	if queue == nil {
+		buildJSONErrorResponse(w, QueueDoesNotExists, http.StatusNotFound)
+		return
+	}
+	if appState != "active" {
+		buildJSONErrorResponse(w, fmt.Sprintf("The provided state is invalid: %s", appState), http.StatusBadRequest)
```
Comment (chia7712) [inline pkg/webservice/handlers.go:930]: Could we align the error message? for example: `"Only following application states are allowed: active"`

**Review by chia7712 — COMMENTED** [review]
@blueBlue0102 thanks for this patch.

File: pkg/webservice/handlers.go [hunk 4]
```diff
@@ -898,6 +898,56 @@ func getPartitionRules(w http.ResponseWriter, r *http.Request) {
...
+func getQueueApplicationsByState(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+	vars := httprouter.ParamsFromContext(r.Context())
+	if vars == nil {
+		buildJSONErrorResponse(w, MissingParamsName, http.StatusBadRequest)
+		return
+	}
+
+	partition := vars.ByName("partition")
+	queueName := vars.ByName("queue")
```
Comment (chia7712) [inline pkg/webservice/handlers.go:910]: Could you please handle the unescaped queue name? see following link for example

https://github.com/apache/yunikorn-core/blob/master/pkg/webservice/handlers.go#L679

Comment (blueBlue0102) [reply pkg/webservice/handlers.go:915]: Done. I added the validation logic starting from line 924.

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (blueBlue0102) [reply pkg/webservice/handlers.go]: Done. Merged the latest master commit and made the adjustment.

File: pkg/webservice/handlers.go (hunk 2 above)
Comment (blueBlue0102) [reply pkg/webservice/handlers.go]: Thanks for the suggestion. The adjustment has been made.

File: pkg/webservice/handlers.go (hunk 3 above)
Comment (blueBlue0102) [reply pkg/webservice/handlers.go]: Good idea. The adjustment has been made.

**Review by chenyulin0719 — APPROVED** [review]
+1 LGTM, built and tested APIs in my env, worked as expected.



- http://localhost:9889/ws/v1/partition/default/queue/root.sandbox/applications/rejected
- http://localhost:9889/ws/v1/partition/default/queue/root.sandbox/applications/active
- http://localhost:9889/ws/v1/partition/default/queue/root.sandbox/applications/active?status=running
- http://localhost:9889/ws/v1/partition/default/queue/root.sandbox/applications/active?status=

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request closed without merging**